### ISSUE TITLE
Class or interface `Aws\Result` does not exist [cache issue]

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#673]
```console
composer require aws/aws-sdk-php:3.54.3
```
```php
<?php
namespace T;

use Aws\Result;
use Aws\S3\S3Client;

class C
{
    /**
     * @var S3Client
     */
    private $client;

    public function __construct()
    {
        $this->client = new S3Client(['region' => 'us-west-2', 'version' => 'latest']);
    }

    public function getObject(string $bucketName, string $key): Result
    {
        return $this->client->getObject([
            'Bucket' => $bucketName,
            'Key' => $key
        ]);
    }
}
```
On the second run I get the following issue reported
```
ERROR: UndefinedClass - src/Q.php:19:65 - Class or interface Aws\Result does not exist
    public function getObject(string $bucketName, string $key): Result
    {
        return $this->client->getObject([
            'Bucket' => $bucketName,
            'Key' => $key
        ]);
    }
```
It's not reported with `--no-cache` or on the first run after `--clear-cache`.

I tried to make the reproduce code self-contained (getting rid of `aws-sdk`), but couldn't reproduce the issue this way.

**Comment by muglug** [comment]
Hmm - I'm unable to reproduce on an empty repo with that single file.

**Comment by weirdan** [comment]
Try this repo: https://github.com/weirdan/psalm-aws-result-bug

```console
$ git clone git@github.com:weirdan/psalm-aws-result-bug.git t2
Cloning into 't2'...
remote: Counting objects: 8, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 8 (delta 0), reused 8 (delta 0), pack-reused 0
Receiving objects: 100% (8/8), done.
$ cd t2
$ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 6 installs, 0 updates, 0 removals
  - Installing mtdowling/jmespath.php (2.4.0): Loading from cache
  - Installing psr/http-message (1.0.1): Loading from cache
  - Installing guzzlehttp/psr7 (1.4.2): Loading from cache
  - Installing guzzlehttp/promises (v1.3.1): Loading from cache
  - Installing guzzlehttp/guzzle (6.3.2): Loading from cache
  - Installing aws/aws-sdk-php (3.54.3): Loading from cache
guzzlehttp/guzzle suggests installing psr/log (Required for using the Log middleware)
aws/aws-sdk-php suggests installing aws/aws-php-sns-message-validator (To validate incoming SNS notifications)
aws/aws-sdk-php suggests installing doctrine/cache (To use the DoctrineCacheAdapter)
Generating autoload files
$ psalm
Scanning files...
Analyzing files...

------------------------------
No errors found!
------------------------------

Checks took 0.21 seconds and used 24.422MB of memory
Psalm was able to infer types for 87.500% of the codebase
$ psalm
Scanning files...
Analyzing files...

ERROR: UndefinedClass - src/Q.php:19:65 - Class or interface Aws\Result does not exist
    public function getObject(string $bucketName, string $key): Result
    {
        return $this->client->getObject([
            'Bucket' => $bucketName,
            'Key' => $key
        ]);
    }


------------------------------
1 errors found
------------------------------

Checks took 0.05 seconds and used 21.845MB of memory
Psalm was able to infer types for 87.500% of the codebase
```

**Comment by weirdan** [comment]
Second run with `--debug` emits the following line:
```
Using reflection to locate file for Aws\result
```
while the first run reports:
```
Using composer to locate file for Aws\Result
Parsing /home/weirdan/src/t2/vendor/aws/aws-sdk-php/src/Result.php
Scanning /home/weirdan/src/t2/vendor/aws/aws-sdk-php/src/Result.php
```
So it might be a case-related issue somewhere.

**Comment by muglug** [comment]
OH I'm running with `composer require --dev vimeo/psalm` and then `vendor/bin/psalm`

**Comment by muglug** [comment]
I get the same with the latest phar

**Comment by weirdan** [comment]
> OH I'm running with composer require --dev vimeo/psalm and then vendor/bin/psalm

Reproducible that way as well, when I run it twice.

**Comment by muglug** [comment]
what OS?

**Comment by weirdan** [comment]
Debian GNU/Linux buster/sid
```console
$ uname -a
Linux portable-ws 4.14.0-3-amd64 #1 SMP Debian 4.14.17-1 (2018-02-14) x86_64 GNU/Linux
```

**Comment by muglug** [comment]
Good ol' case-insensitive Mac OS. OK, will fix.

**Comment by muglug** [comment]
Psalm is confused by `vendor/aws/aws-sdk-php/src/Waiter.php` which lowercases `Result`

**Comment by muglug** [comment]
and Psalm remembers that that lookup failed, and erroneously applies it to uppercased variants. I've seen this bug before, thought it was sorted.
